File: frontend/src/api/predictive.ts
import api from './axios';

export type PredictiveMode = 'off' | 'silent' | 'admin' | 'active';
export type HealthLevel = 'normal' | 'watch' | 'alert' | 'critical' | 'no_data';
export type PredictiveAlertStatus =
  | 'new' | 'in_review' | 'inspection_planned' | 'intervention_required'
  | 'intervention_done' | 'false_positive' | 'monitoring' | 'closed';

export interface PredictiveReason {
  code: string;
  params: Record<string, string | number | null>;
  observed: number | null;
  expected: number | null;
  unit: string | null;
}

export interface PredictiveFactor {
  code: string;
  family: string;
  weight: number;
  value: number;
  quality: number;
  contribution: number;
  observed: number | null;
  expected: number | null;
  unit: string | null;
  params: Record<string, string | number | null>;
  reason: string | null;
}

export interface PredictiveMachineRow {
  equipment_id: string;
  name: string;
  code: string | null;
  department: string | null;
  family: string | null;
  criticality: string | null;
  score: number;
  level: HealthLevel;
  confidence: number | null;
  quality_score: number | null;
  mtbf_pct: number | null;
  maturity: string | null;
  context: string | null;
  ts: string | null;
  open_alerts: number;
}

export interface PredictiveOverview {
  mode: PredictiveMode;
  visible: boolean;
  machines: PredictiveMachineRow[];
  kpis: {
    machines_tracked?: number;
    machines_critical?: number;
    machines_alert?: number;
    machines_insufficient?: number;
    alerts_open?: number;
    alerts_new?: number;
    avg_confidence?: number | null;
    feedback_confirmed_30d?: number;
    feedback_false_positive_30d?: number;
    breakdowns_prevented_30d?: number;
    sensor_problems?: number;
  };
  next_inspections?: {
    alert_id: string; equipment_id: string; due: string; level: string; kind: string;
  }[];
}

export interface PredictiveAlertItem {
  id: string;
  equipment_id: string;
  equipment_name: string | null;
  machine_id: string | null;
  created_at: string | null;
  updated_at: string | null;
  level: HealthLevel;
  score: number;
  kind: string;
  probable_component: string | null;
  probable_failure: string | null;
  reasons: PredictiveReason[];
  sensors_involved: string[];
  window_hours: number | null;
  confidence: number | null;
  recommendation: string | null;
  silent: boolean;
  status: PredictiveAlertStatus;
  assigned_to_id: string | null;
  inspection_due: string | null;
  inspection_result: string | null;
  resolved_at: string | null;
  auto_resolved: boolean;
  ticket_id: string | null;
  engine_version: string | null;
  config_version: number | null;
  feedback?: PredictiveFeedback[];
}

export interface PredictiveFeedback {
  id: string;
  was_correct: boolean | null;
  problem_found: boolean | null;
  component: string | null;
  failure_mode: string | null;
  cause: string | null;
  timing: 'early' | 'on_time' | 'late' | null;
  action_taken: string | null;
  part_replaced: boolean | null;
  prevented_breakdown: boolean | null;
  back_to_normal: boolean | null;
  comments: string | null;
  created_at: string | null;
}

export interface FeedbackInput {
  was_correct?: boolean | null;
  problem_found?: boolean | null;
  component?: string | null;
  failure_mode?: string | null;
  cause?: string | null;
  timing?: 'early' | 'on_time' | 'late' | null;
  action_taken?: string | null;
  part_replaced?: boolean | null;
  prevented_breakdown?: boolean | null;
  back_to_normal?: boolean | null;
  comments?: string | null;
}

export interface MachineHealth {
  mode: PredictiveMode;
  visible: boolean;
  equipment?: {
    id: string; name: string; code: string | null;
    criticality: string | null; family: string | null; department: string | null;
  };
  latest?: {
    ts: string; score: number; level: HealthLevel; context: string | null;
    factors: PredictiveFactor[];
    data_quality: Record<string, { status: string; issues: string[]; last_at: string | null }>;
    quality_score: number | null; confidence: number | null;
    mtbf_pct: number | null; maturity: string | null;
    engine_version: string | null; config_version: number | null;
  } | null;
  history?: { ts: string; score: number; level: HealthLevel; confidence: number | null; mtbf_pct: number | null }[];
  baselines?: {
    metric_key: string; context_key: string; unit: string | null; n_samples: number;
    median: number | null; mad: number | null; mean: number | null; std: number | null;
    p05: number | null; p95: number | null; valid: boolean; frozen: boolean;
    version: number; computed_at: string | null;
  }[];
  alerts?: PredictiveAlertItem[];
  mtbf?: {
    hours_since_last: number; mtbf_hist_h: number | null; mtbf_recent_h: number | null;
    pct_consumed: number | null; trend: number | null; failures_365d: number;
  } | null;
}

export async function fetchPredictiveOverview(): Promise<PredictiveOverview> {
  const { data } = await api.get('/api/predictive/overview');
  return data;
}

export async function fetchMachineHealth(equipmentId: string, hours = 168): Promise<MachineHealth> {
  const { data } = await api.get(`/api/predictive/machines/${equipmentId}`, { params: { hours } });
  return data;
}

export async function fetchPredictiveAlerts(params: {
  status?: string; level?: string; open_only?: boolean; equipment_id?: string;
  limit?: number; offset?: number;
} = {}): Promise<{ total: number; items: PredictiveAlertItem[]; mode: PredictiveMode; visible: boolean }> {
  const { data } = await api.get('/api/predictive/alerts', { params });
  return data;
}

export async function updateAlertStatus(
  alertId: string,
  body: { status: PredictiveAlertStatus; assigned_to_id?: string; inspection_due?: string; inspection_result?: string },
): Promise<PredictiveAlertItem> {
  const { data } = await api.patch(`/api/predictive/alerts/${alertId}/status`, body);
  return data;
}

export async function submitAlertFeedback(alertId: string, body: FeedbackInput): Promise<{ id: string }> {
  const { data } = await api.post(`/api/predictive/alerts/${alertId}/feedback`, body);
  return data;
}

export async function createAlertTicket(alertId: string): Promise<{ ticket_id: string; ticket_number: string }> {
  const { data } = await api.post(`/api/predictive/alerts/${alertId}/ticket`);
  return data;
}

export async function evaluateNow(equipmentId: string): Promise<unknown> {
  const { data } = await api.post(`/api/predictive/machines/${equipmentId}/evaluate`);
  return data;
}

export async function fetchPredictiveSettings(): Promise<{
  plant: Record<string, unknown> & { mode: PredictiveMode };
  machines: { equipment_id: string; enabled: boolean | null; mode: PredictiveMode | null }[];
}> {
  const { data } = await api.get('/api/predictive/settings');
  return data;
}

export async function updatePredictiveSettings(body: Record<string, unknown>): Promise<{ ok: boolean; version: number }> {
  const { data } = await api.put('/api/predictive/settings', body);
  return data;
}

export async function updateMachinePredictiveSettings(
  equipmentId: string,
  body: { enabled?: boolean | null; mode?: PredictiveMode | null },
): Promise<{ ok: boolean }> {
  const { data } = await api.put(`/api/predictive/machines/${equipmentId}/settings`, body);
  return data;
}

export async function runBacktest(body: {
  equipment_id: string; start: string; end: string; step_min?: number;
}): Promise<{
  metrics: {
    evaluations: number; alerts: number; failures: number; detected: number;
    missed: number; false_positives: number; avg_lead_hours: number | null;
  };
  snapshots: { ts: string; score: number; level: HealthLevel }[];
  alerts: { opened_at: string; kind: string; level: string; score: number }[];
  failures: { started_at: string; detected: boolean; lead_hours: number | null }[];
}> {
  const { data } = await api.post('/api/predictive/backtest', body);
  return data;
}

File: backend/app/schemas/predictive.py
"""Pydantic schemas for the predictive intelligence API."""
from __future__ import annotations

from datetime import datetime
from typing import Any, Optional
from uuid import UUID

from pydantic import BaseModel, Field


class SettingsUpdate(BaseModel):
    mode: Optional[str] = Field(None, pattern="^(off|silent|admin|active)$")
    eval_interval_min: Optional[int] = Field(None, ge=5, le=1440)
    baseline_refresh_hours: Optional[int] = Field(None, ge=1, le=168)
    weights: Optional[dict[str, float]] = None
    levels: Optional[dict[str, float]] = None
    windows: Optional[dict[str, float]] = None
    persistence_evals: Optional[int] = Field(None, ge=1, le=10)
    cooldown_hours: Optional[float] = Field(None, ge=0, le=168)
    confidence_floor: Optional[float] = Field(None, ge=0, le=1)
    fingerprint_windows_h: Optional[list[float]] = None
    prefailure_exclude_h: Optional[int] = Field(None, ge=1, le=168)
    baseline_window_days: Optional[int] = Field(None, ge=7, le=180)
    baseline_min_samples: Optional[int] = Field(None, ge=5, le=5000)
    baseline_drift_cap_pct: Optional[float] = Field(None, ge=1, le=500)


class MachineSettingsUpdate(BaseModel):
    enabled: Optional[bool] = None
    mode: Optional[str] = Field(None, pattern="^(off|silent|admin|active)$")
    overrides: Optional[dict[str, Any]] = None


class RuleCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=200)
    equipment_id: Optional[UUID] = None
    metric_key: str = Field(..., min_length=1, max_length=120)
    aggregation: str = Field("avg", pattern="^(avg|max|min|last|slope|count)$")
    window_hours: float = Field(1.0, gt=0, le=168)
    operator: str = Field("gt", pattern="^(gt|lt)$")
    threshold: float
    severity: str = Field("alert", pattern="^(watch|alert|critical)$")
    enabled: bool = True


class RuleUpdate(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=200)
    metric_key: Optional[str] = None
    aggregation: Optional[str] = Field(None, pattern="^(avg|max|min|last|slope|count)$")
    window_hours: Optional[float] = Field(None, gt=0, le=168)
    operator: Optional[str] = Field(None, pattern="^(gt|lt)$")
    threshold: Optional[float] = None
    severity: Optional[str] = Field(None, pattern="^(watch|alert|critical)$")
    enabled: Optional[bool] = None


class AlertStatusUpdate(BaseModel):
    status: str = Field(..., pattern="^(new|in_review|inspection_planned|intervention_required|intervention_done|false_positive|monitoring|closed)$")
    assigned_to_id: Optional[UUID] = None
    inspection_due: Optional[datetime] = None
    inspection_result: Optional[str] = None


class FeedbackCreate(BaseModel):
    was_correct: Optional[bool] = None
    problem_found: Optional[bool] = None
    component: Optional[str] = Field(None, max_length=200)
    failure_mode: Optional[str] = Field(None, max_length=200)
    cause: Optional[str] = Field(None, max_length=500)
    timing: Optional[str] = Field(None, pattern="^(early|on_time|late)$")
    action_taken: Optional[str] = Field(None, max_length=500)
    part_replaced: Optional[bool] = None
    prevented_breakdown: Optional[bool] = None
    back_to_normal: Optional[bool] = None
    comments: Optional[str] = None


class BacktestRequest(BaseModel):
    equipment_id: UUID
    start: datetime
    end: datetime
    step_min: int = Field(60, ge=15, le=1440)


class FailureConfirm(BaseModel):
    confirmed: bool = True
    component: Optional[str] = Field(None, max_length=200)
    failure_type: Optional[str] = Field(None, max_length=200)
    notes: Optional[str] = None
